More realistic gas explosions for bloated and smoker

Coolthulhu opened the pull request:
Addresses the most common cause of #14894

Exploding bloated and smokers will now use a gas propagation function. In open field, it doesn't differ from what happens in master (except that it calculates slower).

When constrained by walls, it will propagate more in the opposite direction.

When constrained by fields of the same type, it will propagate through them, losing some power in the process.

mugling commented:
Interestingly I was working on an implementation of this yesterday but you've made much more progress. I was wanting to add an `emit` field to `itype` so we can have proper smoke grenades. How much work would it be to extend your PR to support this as it looks like most of the groundwork is inplace?

Coolthulhu commented:
Mostly just cleaning up would be required.
If the function is to be used more globally, it needs to be in an accessible file/class, such as `map` or `explosion.cpp`. Having it in `map` would also allow making it less slow (the implementation is rather naive).

mugling commented:
Do you want to try for that approach - indeed the function you wrote has an identical signature to the `map::emit_field` that I was working on

Coolthulhu commented:
Moved to field.cpp

Currently doesn't support non-gases properly, because it has a hardcoded check for `TFLAG_PERMEABLE`, but that's easy to fix - just needs the field type to specify whether it is a gas or not.

Coolthulhu commented:
I added a phase field to fields. Changed the propagation function to respect that by not propagating non-gases through permeable but impassable tiles.

Added propagation to boomer bile on its death as an example.

mugling commented on src/enums.h:
```diff
@@ -160,7 +160,7 @@ enum artifact_natural_property {
 };
 
-enum phase_id {
+enum phase_id : int {
```
`enum class` avoids the need for that ugly untyped `PNULL`

mugling commented on src/mondeath.cpp:
```diff
@@ -136,7 +137,7 @@ void mdeath::normal(monster *z)
             }
 
             if( tarp != tripoint_min ) {
-                g->m.spawn_item( tarp, meat );
```
Unrelated changes?

Coolthulhu replied:
Kinda related to critters exploding, though not to propagation
